src/treq/auth.py

```diff
@@ -1,13 +1,223 @@
+        :param qop: Quality Of Protection HTTP Digest auth param
+        :param opaque: "opaque" HTTP Digest authentication param
+             (should be sent back to server unchanged)
+        :param cached: Identifies that authentication already have been
+             performed for URI/method,
+             and new request should use the same params as first
+             authenticated request
+        :param path: the URI path where we are authenticating
+        :param method: HTTP method to be used when requesting
+
+        :return: HTTP Digest authentication string
+        """
+        algo = algorithm.upper()
+        original_algo = algorithm
+        path_parsed = urlparse(path)
+        actual_path = path_parsed.path
+
+        if path_parsed.query:
+            actual_path += b'?' + path_parsed.query
+
+        a1 = self._username
+        a1 += b':'
+        a1 += realm
+        a1 += b':'
+        a1 += self._password
+
+        a2 = method
+        a2 += b':'
+        a2 += actual_path
+
+        if algo == b'MD5' or algo == b'MD5-SESS':
+            digest_hash_func = _md5_utf_digest
+        elif algo == b'SHA':
+            digest_hash_func = _sha1_utf_digest
+        elif algo == b'SHA-256':
+            digest_hash_func = _sha256_utf_digest
+        elif algo == b'SHA-512':
+            digest_hash_func = _sha512_utf_digest
+        else:
+            raise UnknownDigestAuthAlgorithm(algo)
+
+        ha1 = digest_hash_func(a1)
+        ha2 = digest_hash_func(a2)
+
+        cnonce = _generate_client_nonce(nonce)
+
+        if algo == b'MD5-SESS':
+            sess = ha1.encode('utf-8')
+            sess += b':'
+            sess += nonce
+            sess += b':'
+            sess += cnonce.encode('utf-8')
+            ha1 = digest_hash_func(sess)
+
+        if cached:
+            self._digest_auth_cache[(method, path)]['c'] += 1
+            nonce_count = self._digest_auth_cache[(method, path)]['c']
+        else:
+            nonce_count = 1
+
+        ncvalue = '%08x' % nonce_count
+        if qop is None:
+            rd = ha1.encode('utf-8')
+            rd += b':'
+            rd += ha2.encode('utf-8')
+            rd += b':'
+            rd += nonce
+            response_digest = digest_hash_func(rd).encode('utf-8')
+        else:
+            rd = ha1.encode('utf-8')
+            rd += b':'
+            rd += nonce
+            rd += b':'
+            rd += ncvalue.encode('utf-8')
+            rd += b':'
+            rd += cnonce.encode('utf-8')
+            rd += b':'
+            if not isinstance(qop, bytes):
+                rd += qop.encode('utf-8')
+            else:
+                rd += qop
+            rd += b':'
+            rd += ha2.encode('utf-8')
+            response_digest = digest_hash_func(rd).encode('utf-8')
+        hb = b'username="'
+        hb += self._username
+        hb += b'", realm="'
+        hb += realm
+        hb += b'", nonce="'
```

Comment:
This repeated pattern here, with no quoting, makes me think that we are just begging for someone to put in a header-splitting attack here.  Is this implementation robust against bad data for some other reason, or do we have some reason to believe that all the parameters here are always trusted?

Reply:
> This repeated pattern here, with no quoting, makes me think that we are just begging for someone to put in a header-splitting attack here. Is this implementation robust against bad data for some other reason, or do we have some reason to believe that all the parameters here are always trusted?

Hmm, is there infrastructure for generating headers in treq/twisted already that might be of use here? I suppose it could allow a header-splitting attack in theory, although I haven't used this code myself for anything where that would likely be a viable attack.

Reply:
Well I went ahead and just refactored it to function like the [latest code in requests](https://github.com/psf/requests/blob/v2.28.1/requests/auth.py#L126-L234) now that we don't need python2 compatibility so it shouldn't be any worse than that security wise I think at least.